### PR TITLE
Reduce lock usage during EditDistance.GetEditDistance and BKTree.Find

### PR DESCRIPTION
**Pull request by ToddGrun**
The SimplePool and ArrayPool classes previously always locked on Allocate/Free which is showing up in a cpu usage in a trace David Kean referred to us. Instead, as the char arrays being allocated are typically quite small, utilize the stack for these temporary char array locations when possible, falling back to heap allocating in the extremely rare occasions where it's not.

Below is the cpu usage that David pointed us to:
image

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/BKTree.cs [hunk 1]
```diff
@@ -70,25 +70,20 @@ public void Find(ref TemporaryArray<string> result, string value, int? threshold
...
             return;
 
-        var lowerCaseCharacters = ArrayPool<char>.GetArray(value.Length);
-        try
-        {
-            for (var i = 0; i < value.Length; i++)
-                lowerCaseCharacters[i] = CaseInsensitiveComparison.ToLower(value[i]);
+        Span<char> lowerCaseCharacters = value.Length < 512
+            ? stackalloc char[value.Length]
+            : new char[value.Length];
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/BKTree.cs:75]: I would still use the pool for large strings.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/BKTree.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/BKTree.cs:75]: Just here, or in EditDistance too? I'm not sure about here, but I wouldn't think the EditDistance code would almost ever exceed 512 chars.

Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/BKTree.cs:75]: Just to make sure, I don't think the old code was pooling large strings, as ArrayPool checked against MaxPooledArraySize for both retrieving and returning from the pool.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/BKTree.cs:75]: Ah. Never mind :-)

**Pull request merged**